osbrain/tests/test_agent_sync_publications.py

```diff
@@ -73,6 +74,152 @@ def on_error(agent):
+        """
+        if len(agent.received) < 2:
+            return False
+        return agent.received[-1] - agent.received[-2] != 1
+
+    def publish(agent):
+        assert agent.count % 2 == 1
+        agent.send('pub', agent.count, topic='odd')
+        agent.send('pub', agent.count + 1, topic='even')
+        agent.count += 2
+
+    # Set up the agents
+    server = run_agent('server')
+    client = run_agent('client', base=Client)
+
+    server.set_attr(count=1)
+    addr = server.bind(socket_type, alias='pub', handler=lambda: None)
+    client.connect(addr, alias='sub', handler={'odd': append_received,
+                                               'even': append_received})
+
+    # Make sure the connection is stablished
+    server.each(0.1, 'send', 'pub', 'connecting...', topic='odd', alias='tmp')
+    assert wait_agent_attr(client, data='connecting...')
+    server.stop_timer('tmp')
+
+    # Wait until client receives some numbers
+    client.set_attr(received=[])
+    server.each(0.1, publish)
+    assert wait_agent_attr(client, length=5)
+
+    # Unsubscribe
+    client.unsubscribe('sub', 'odd')
+
+    assert wait_agent_condition(client, check_non_consecutive)
+    assert client.get_attr('received')[-1] % 2 == 0
+
+
+@pytest.mark.parametrize('socket_type', ['PUB', 'SYNC_PUB'])
+@pytest.mark.parametrize('subscribe_separately', [True, False])
+def test_subscribe(nsproxy, socket_type, subscribe_separately):
+    '''
+    Test subscribing to various topics/handlers works fine.
+    '''
+    server = run_agent('server')
+    client = run_agent('client')
+
+    addr = server.bind(socket_type, alias='pub', handler=lambda: None)
+    client.set_attr(received=[])
+    client.connect(addr, alias='sub', handler=append_received)
+
+    # Stablish a connection
+    server.each(0.1, 'send', 'pub', 'connecting...', alias='tmp')
+    assert wait_agent_attr(client, data='connecting...')
+    server.stop_timer('tmp')
+
+    # Subscribe to two topics
+    client.unsubscribe('sub', '')
+    if subscribe_separately:
+        client.subscribe('sub', handlers={'negate': receive_negate})
+        client.subscribe('sub', handlers={'normal': append_received})
+    else:
+        client.subscribe('sub', handlers={'negate': receive_negate,
+                                          'normal': append_received})
+
+    # Only message from subscribed topic received
+    server.each(0.1, 'send', 'pub', 1, topic='negate')
+    server.each(0.1, 'send', 'pub', 2, topic='normal')
+    server.each(0.1, 'send', 'pub', 'not received', topic='topic_C')
+
+    assert wait_agent_attr(client, data=-1)
+    assert wait_agent_attr(client, data=2)
+    assert not wait_agent_attr(client, data=1, timeout=0.5)
+    assert not wait_agent_attr(client, data=-2, timeout=0.0)
+    assert not wait_agent_attr(client, data='not received', timeout=0.0)
+
+
+@pytest.mark.parametrize('socket_type', ['PUB', 'SYNC_PUB'])
+def test_resubscribe(nsproxy, socket_type):
```

Comment:
Not sure if this `test_resubscribe` is an interesting/required test. Even if so, I would call it `test_unsubscribe_and_subscribe_again`.

For me, a `test_resubscribe` would be one that actually checked the behavior in this case:

```python
client.subscribe('sub', handler={'x': foo})
client.subscribe('sub', handler={'x': bar})
```

Subscription after subscription, without an `unsubscribe` call in between.

This is currently not defined behavior. In my opinion we could simply override the first subscription. What you think?

Reply:
I agree with everything you said. I renamed the test, and I will work now on writing the true ``test_resubscribe``. I will comment again when I'm done with it. 

For me, I would leave the ``test_unsubscribe_and_subscribe_again`` in the test code base, as it won't hurt to have more tests. What do you think?

Reply:
Implemented the new test - see commit 9d44e10.

Reply:
Yeah, we can leave it there. :blush: